### ISSUE TITLE
`dependent: :destroy` on model, doesn't propagating the children destroy call return value

### ISSUE THREAD

**Issue by kelvinst** [rails/rails#17056]
So, I have the following models and relations configured:

``` ruby
class Board < ActiveRecord::Base
  has_many :lists, dependent: :destroy
end

class List < ActiveRecord::Base
  has_many :cards, dependent: :restrict_with_error
end

class Card < ActiveRecord::Base
  belongs_to :list
end
```

With this, I expect that when I try to delete a `Board` which have a `List` that have a `Card` on it, all the process of cascade deletion is cancelled, and neither board, lists or cards are deleted.

But the actual behavior of this is:
- The list deletion is cancelled because of `dependent: :restrict_with_error`
- It doesn't affect the board deletion, and the board is deleted leaving some orphans lists on db

To workaround this, I can set the `dependent: :restrict_with_exception` and, as on the `List` deletion a execption will be raised, the whole transaction is rollbacked. But this is not ideal for me, because in this way, I must to threat with an exception that's not expected, given I am calling the method `@board.destroy`, unbanged.

There is somebody that has solved this problem in some decent way? Or maybe it can be considered a bug of `dependent: :destroy` do not propagate the associated destroy return value?

**Comment by NirvashPrime** [comment]
I see this hasn't been updated in several years, is this still an issue?

**Comment by jameswangz** [comment]
I am still encountering this issue, my rails version is 5.1.4.

**Comment by why-el** [comment]
This is no longer a problem in master, evidence below. So I think this should be closed, cc @kelvinst.

@jamesarosen I think the logical thing is happening here. The restriction happens on the right object and thus you see the error there. It might not always be possible to even have the parent record initiated so it makes little sense to try and add an error there (it might be lazily loaded for instance). 

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :boards do |t|
  end

  create_table :lists do |t|
    t.references :board
  end

  create_table :cards do |t|
    t.references :list
  end
end

class Board < ActiveRecord::Base
  has_many :lists, dependent: :destroy
end

class List < ActiveRecord::Base
  has_many :cards, dependent: :restrict_with_error
end

class Card < ActiveRecord::Base
  belongs_to :list
end

class BugTest < Minitest::Test
  def test_cascading_deletion_cancellation
    board = Board.create
    list = board.lists.create
    list.cards.create

    assert_equal 1, Board.count
    assert_equal 1, List.count
    assert_equal 1, Card.count

    board.destroy

    assert_equal 1, Card.count
    assert_equal 1, List.count
    assert_equal 1, Board.count
  end
end
```

**Comment by jonathanhefner** [comment]
Thank you @why-el for the reproduction script!  I can confirm this is fixed in both Rails 5.2.4.2 and Rails 6.0.3.
